backend/src/indexer/block_preprocessor.rs

```diff
@@ -281,49 +300,105 @@ impl concordium_rust_sdk::indexer::Indexer for BlockPreProcessor {
-/// prior to protocol version 4, in this case the account info is used to get
-/// the stake for the validator and the total pool for the validator prior to
-/// protocol version 4 will be set to their account stake.
+/// Responsible for computing the latest validator staking information which
+/// returns a Tuple of <Amount, ValidatorStakingInformation> - where Amount is
+/// the total stake summed for all the validators
+/// and the ValidatorStakingInformation is a struct containing all of the new
+/// information we need to update for the validators such as their staked
+/// amounts and their pool staked amounts. As it is intensive to go to the node
+/// through the client, the last computed total stake is stored in a cache. We
+/// only go to the retrieve the latest stakes if enough blocks have passed
+/// (defined by our env variable STAKE_RECOMPUTE_INTERVAL_IN_BLOCKS). It is
+/// important to note that in our cached scenario, we just return the last
+/// cached total stake amount - we do not return validatorStakingInformation
+/// because: if this is present when we reach the block processor it will
+/// perform database update queries for all the validator information provided,
+/// therefore it is crucial to return a struct that represents 'no updates at
+/// this time'
 pub async fn compute_validator_staking_information(
     client: &mut v2::Client,
     block_height: v2::BlockIdentifier,
 ) -> v2::QueryResult<(Amount, ValidatorStakingInformation)> {
-    let mut total_staked = Amount::zero();
-    let mut bakers = client.get_baker_list(block_height).await?.response;
+    let current_block_height: u64 = get_block_height_from_absolute(block_height);
+    let cache_read = TOTAL_STAKED_AMOUNTS_CACHE.read().await;
+    let computed_stake_result: v2::QueryResult<(Amount, ValidatorStakingInformation)>;
+    let mut should_update_cache = false;
+
+    // Read from the cache, check if its time to recompute the staked amounts or if
+    // we can just return the cached total staked amount
+    if let Some(cache) = &*cache_read {
+        // check if enough blocks have passed to recompute stakes
+        if current_block_height - cache.block_height == *STAKE_RECOMPUTE_INTERVAL_IN_BLOCKS {
+            computed_stake_result =
+                compute_validator_staking_information_with_client(client, block_height).await;
+            should_update_cache = true;
+        } else {
+            // return cached values with no updates required staking information
+            let no_updates_required_for_validators = ValidatorStakingInformation {
+                ids: vec![],
+                staked_amounts: vec![],
+                pool_total_staked_amounts: vec![],
+            };
+            debug!(
+                "Returning cache total stake for height: {}, total staked: {}",
+                current_block_height, cache.total_staked
+            );
+            computed_stake_result = Ok((cache.total_staked, no_updates_required_for_validators));
```

Comment:
if we return/leave the `compute_validator_staking_information` function here, then we don't need to specify/mark `should_update_cache` as true in all other arms of this conditional statement. In general I find it a confusing flow to have a default ` let mut should_update_cache = false;` specified at the top and then except for one arm to update its value to `true`. Would be good to get rid-off this `should_update_cache` variable.